### PR TITLE
Rahul

### PR DESCRIPTION
**Pull request by rahulk789**
clone this branch and compile it. If its compiled just merge it with the main branch . i cannot check the functioning of the class functions yet because we have Nextleveldb class remaining

have a nice day bro

### CODE REVIEW

File: src/nextleveldb/Document.java [hunk 1]
```diff
@@ -5,10 +5,27 @@
 
 class Document {
 	String name;
-	private Map<String, String> pairs = new HashMap<String, String>();
+	private Map<String,Map<String,String>>pairs = new HashMap<String,Map<String,String>>();
```
Comment (snprajwal) [inline src/nextleveldb/Document.java]: A pair is a key mapped to value, both are strings. Why are we mapping a string to another map?

File: src/nextleveldb/Document.java [hunk 2]
```diff
@@ -5,10 +5,27 @@
...
 
 	Document(String name) {
 		this.name = name;
 	}
-	// add functions for create, delete, list, update pairs
+
+	void listPairs(){
+		for(Map.Entry pairEntry: pairs.entrySet()) {
```
Comment (snprajwal) [inline src/nextleveldb/Document.java]: Can do this in a better way with lambdas, something like
`map.forEach( (k,v) -> System.out.println("Key: " + k + ": Value: " + v));`

File: src/nextleveldb/Document.java [hunk 3]
```diff
@@ -5,10 +5,27 @@
...
+	void createPair(String keyy , String val){
+		pairs.get(this.name).put(keyy,val);
+	}
+	void updatePair(String keyy , String val){
+		pairs.get(this.name).replace(keyy,val);
+	}
+	void deletePair(String keyy){
+		pairs.get(this.name).remove(keyy);
+	}
+
```
Comment (snprajwal) [inline src/nextleveldb/Document.java]: why are we getting the doc name from the map? :thinking: it'll return a string
need to rethink the logic being used here, doesn't look right

File: src/nextleveldb/Index.java [hunk 4]
```diff
@@ -5,10 +5,28 @@
 
 class Index {
 	String name;
-	private Map<String, Document> docs = new HashMap<String, Document>();
+	String currentind;
+	Index currentIndex;
```
Comment (snprajwal) [inline src/nextleveldb/Index.java]: don't need these two, we're already maintaining a ref in the parser

File: src/nextleveldb/Index.java [hunk 5]
```diff
@@ -5,10 +5,28 @@
...
 
 	Index(String name) {
-		this.name = name;
+		this.name=name;
+	}
+	Document getDocument(String name){
+		if (!docs.containsValue(name)){
+			return new Document("");}
+		else{
+			return new Document(name);
```
Comment (snprajwal) [inline src/nextleveldb/Index.java]: we don't wanna return a new, we wanna return ref to the existing one, with `docs.get(name)`

File: src/nextleveldb/Index.java [hunk 6]
```diff
@@ -5,10 +5,28 @@
...
+	Document getDocument(String name){
+		if (!docs.containsValue(name)){
+			return new Document("");}
+		else{
+			return new Document(name); 
+		}
+	}
+	void listDocuments(){
+		for (Map.Entry pairEntry: docs.entrySet()) {
+			System.out.println(pairEntry.getKey()+":"+pairEntry.getValue());}
```
Comment (snprajwal) [inline src/nextleveldb/Index.java]: shouldn't be printing values, only keys

File: src/nextleveldb/Index.java [hunk 7]
```diff
@@ -5,10 +5,28 @@
...
+	}
+	void createDocument(String name){
+		docs.put(name,getDocument(name));
```
Comment (snprajwal) [inline src/nextleveldb/Index.java]: here we create new :joy: if we create new in the get function, then everytime we do a search, we create a new doc instead of using the existing one

File: src/nextleveldb/Parser.java [hunk 8]
```diff
@@ -1,10 +1,12 @@
 package nextleveldb;
 
+
 class Parser {
 	NextLevelDB db = new NextLevelDB();
+	Index ind = new Index("");
+	Document doc = new Document("");
```
Comment (snprajwal) [inline src/nextleveldb/Parser.java]: why new vars? the two vars below this serve the exact same purpose. Also, we don't store anything with `new` in the `currentIndex` or `currentDocument` vars, they only act as pointers to the objects stored in the maps.

File: src/nextleveldb/Parser.java [hunk 9]
```diff
@@ -56,31 +59,124 @@ void indexHandler(String[] tokens) {
 		case "connect":
 			index = db.getIndex(name);
 			if (!index.name.equals("")) {
+				if (currentIndex.name.equals(name)) {
+					currentIndex = null;
```
Comment (snprajwal) [inline src/nextleveldb/Parser.java]: we don't wanna do this, we just wanna print a message saying "Already connected". Even if we don't do anything, it'll still be fine, we're just reassigning the reference

File: src/nextleveldb/Parser.java [hunk 10]
```diff
@@ -56,31 +59,124 @@ void indexHandler(String[] tokens) {
...
 			System.out.println("ERR: Index does not exist");
 			return;
-		}
-	}
 
+	}	
 	void documentHandler(String[] tokens) {
+		String operation = tokens[0];
+		if (operation.equals("list")) {
+			ind.listDocuments();
```
Comment (snprajwal) [inline src/nextleveldb/Parser.java]: use the currentIndex variable, here and everywhere else. No need of new vars

File: src/nextleveldb/Parser.java [hunk 11]
```diff
@@ -56,31 +59,124 @@ void indexHandler(String[] tokens) {
...
+		if (operation.equals("list")) {
+			doc.listPairs();
+			return;
+		}
+		if (tokens.length < 3) {
+			System.out.println("ERR: Insufficient arguments for operation " + operation.toUpperCase());
+			return;
+		}
+		String name = tokens[2];
+		String keyy = tokens[2];
```
Comment (snprajwal) [inline src/nextleveldb/Parser.java]: rename please :) make it `key`

File: src/nextleveldb/Parser.java [hunk 12]
```diff
@@ -56,31 +59,124 @@ void indexHandler(String[] tokens) {
...
+		String val = tokens[3];
+		Document document;
```
Comment (snprajwal) [inline src/nextleveldb/Parser.java]: use the currentDocument variable from the `parse` function here, no need to create a new one

File: src/nextleveldb/Parser.java [hunk 13]
```diff
@@ -56,31 +59,124 @@ void indexHandler(String[] tokens) {
...
+		switch (operation) {
+		case "create":
+			document=ind.getDocument(name);
```
Comment (snprajwal) [inline src/nextleveldb/Parser.java]: see above comment, we are not passing doc name when creating pair, gotta use the reference from calling func

**Review by snprajwal — COMMENTED** [review]
Left a bunch of comments, check them out. Also, please test your changes locally before committing. Even though index funcs are not done, you can test the doc and pair functions by connecting to a dummy index. It's really hard to test all edge cases since I haven't written this code, counting on you to do that before the next push :)

File: src/nextleveldb/Parser.java (hunk 9 above)
Comment (rahulk789) [reply src/nextleveldb/Parser.java]: didnt intend to add that there , my bad lol

File: src/nextleveldb/NextLevelDB.java [hunk 14]
```diff
@@ -7,15 +7,22 @@
...
-
+        if (!indexes.containsValue(name)){
+            return new Index("");}
+        else{
+            return indexes.get(name); 
+        }
+    }
 	void createIndex(String name) {
-	}
+        indexes.put(name,getIndex(name));
```
Comment (snprajwal) [inline src/nextleveldb/NextLevelDB.java]: Smart! I like it :bird:

File: src/nextleveldb/Parser.java [hunk 15]
```diff
@@ -94,8 +91,9 @@ void documentHandler(String[] tokens) {
 			return;
 		}
 		String name = tokens[2];
-		Document document;
-		switch (operation) {
+		Document document ;
+        document=currentDocument
```
Comment (snprajwal) [inline src/nextleveldb/Parser.java]: No need of duplicate reference here, we can just use `currentDocument` directly wherever we are using `document`

File: src/nextleveldb/Parser.java [hunk 16]
```diff
@@ -147,6 +145,7 @@ void pairHandler(String[] tokens) {
 		String key = tokens[2];
 		String val = tokens[3];
 		Document document;
+        document=currentDocument
```
Comment (snprajwal) [inline src/nextleveldb/Parser.java]: Same here

**Review by snprajwal — COMMENTED** [review]
Awesome! If all of this is working as expected, let me know and I'll merge with main. Can also notify @aum123456 to start work once it is merged
